Спринт 15. Чистая архитектура

svetSanna opened the pull request:
Здравствуйте! Для меня это очень тяжелый спринт. Потратила очень много времени. Буду очень благодарна за полезные развернутые советы, особенно, если что-то нужно переделать. Спасибо!

svetSanna committed:
Спринт 15. Чистая архитектура
15.1. Созданы папки для слоев, почти все файлы разнесены по папкам
15.2. В SettingsActivity работа с SharedPreferences сделана через Creator
15.3. В MediaActivity работа с MediaPlayer сделана в правилах чистой архитектуры

svetSanna committed:
15.4. В MediaActivity работу с локалью убрала в presentation.mapper (SimpleDateFormetMapper)

svetSanna committed:
15.4.1. Еще подправила немного (дополнительно к 15.4)

svetSanna committed:
15.5. Сделала, чтобы при нажатии на кнопку "Медиа", программа не вылетала: теперь пишется сообщение, что посмотрет работу плейера можно через экран Поиска.
(Пока логика работы при переходе на экран медиаплейера и главного экрана нигде не описывалась)

svetSanna committed:
15.6. Сделала выполнение поискового запроса в соответствии с чистой архитктурой

svetSanna committed:
15.7. 1. Исправила отображение ProgressBar при нажатии на кнопку "Очистить строку ввода"
2. Сделала обновление поиска при нажатии на кнопку "Обновить" (после ошибки запроса)

svetSanna committed:
Спринт 15.8. 1. Перенесла SearchHistory в слой data.
2. Убрала из SearchHistory работу с адаптером в ui.SearchActivity (туда, где и используется SearchHistory)

svetSanna committed:
Спринт 15.8. 2. В SearchActivity получаю SharedPreferences через Creator

svetSanna committed:
Спринт 15.9 В SearchActivity сделала работу с историей поиска через Интерактор. Таким образом, в SearchActivity нет ссылок на data-слой.
Хочу еще убрать логику поиска из SearchActivity

svetSanna committed:
Спринт 15.10 Исправлена ошибочка в SearchHistoryInteractor.addItem()

svetSanna committed:
Спринт 15.11 Ctrl+Alt+l

i-masloed commented on app/src/main/java/com/example/playlistmaker/creator/Creator.kt:
```diff
@@ -0,0 +1,65 @@
+    */
+    private lateinit var application: AppCompatActivity
+
+    fun initApplication(application: AppCompatActivity) { // передаем сюда контекст приложения
```
Сюда надо передавать контекст приложения, а не активити

i-masloed commented on app/src/main/java/com/example/playlistmaker/data/repository/TrackRepositoryImpl.kt:
```diff
@@ -0,0 +1,46 @@
+            }
+            return Resource.Success(results) //results
+        } else {
+            return Resource.Error("Произошла сетевая ошибка") //emptyList()
```
строки нужно вынести в ресурсы

i-masloed commented on app/src/main/java/com/example/playlistmaker/domain/entity/Track.kt:
```diff
@@ -19,20 +18,20 @@ data class Track(
-    ) {
-    }
-    fun getCoverArtwork() = artworkUrl100.replaceAfterLast('/',"512x512bb.jpg")
+    /* constructor(parcel: Parcel) : this(
```
Неиспользуемый код стоит удалять

i-masloed commented on app/src/main/java/com/example/playlistmaker/domain/repository/MediaPlayerRepository.kt:
```diff
@@ -0,0 +1,7 @@
+package com.example.playlistmaker.domain.repository
+
+import android.media.MediaPlayer
```
зависимость на Android sdk в доменном слое

i-masloed commented on app/src/main/java/com/example/playlistmaker/domain/use_case/GetMediaPlayerUseCase.kt:
```diff
@@ -0,0 +1,12 @@
+package com.example.playlistmaker.domain.use_case
+
+import android.media.MediaPlayer
```
Зависимость на Android sdk в доменном слое

i-masloed commented on app/src/main/java/com/example/playlistmaker/domain/use_case/GetTrackListUseCase.kt:
```diff
@@ -0,0 +1,56 @@
+                }
+
+                is Resource.Error -> {
+                    consumer.consume(ConsumerData.Error("Ошибка поиска"))
```
Строки не вынесены в ресурсы

i-masloed commented on app/src/main/java/com/example/playlistmaker/ui/MediaActivity/MediaActivity.kt:
```diff
@@ -1,29 +1,36 @@
-    private var mediaPlayer = MediaPlayer()
+    //private var mediaPlayer = MediaPlayer()
+    private var getMediaPlayerUseCase = Creator.provideGetMediaPlayerUseCase()
+    private var mediaPlayer = getMediaPlayerUseCase()
```
Вся работа с медиаплейером должна остаться в дата слое

i-masloed commented on app/src/main/java/com/example/playlistmaker/ui/SettingsActivity/SettingsActivity.kt:
```diff
@@ -1,53 +1,63 @@
-        val sharedPrefs = getSharedPreferences(PLAYLISTMAKER_PREFERENCES, MODE_PRIVATE)
+        //val sharedPrefs = getSharedPreferences(PLAYLISTMAKER_PREFERENCES, MODE_PRIVATE)
+        //Creator.initApplication(this.application)
+        Creator.initApplication(this)
```
Не нужно это вызывать из активити.

i-masloed commented on app/src/main/java/com/example/playlistmaker/ui/SettingsActivity/SettingsActivity.kt:
```diff
@@ -1,53 +1,63 @@
         val selectorSwitch = findViewById<SwitchCompat>(R.id.selector_switch)
         selectorSwitch.setOnCheckedChangeListener { _, checked ->
             (applicationContext as App).switchTheme(checked)
             sharedPrefs.edit()
```
Работу с SharedPreferences нужно вынести в дата слой

svetSanna committed:
Спринт 15.12. После ревью.
1. Работа с MediaPlayer полностью вынесена в data-слой. Для доступа к view-элементам в интерастор передается activity.
2. Некоторые строки добавлены в ресурсы.

svetSanna committed:
Спринт 15.13. Убрала Creator.initApplication в App из SearchActivity и SettingsActivity. Заменила сигнатуру метода initApplication на application:Application? соответственно изменила тиа поля application

svetSanna committed:
Спринт 15.14. Убрала SearchActivity прямое использование SharedPreference. Теперь используется через  интерактор.

svetSanna committed:
Спринт 15.14. Убрала из SettingsActivity прямое использование SharedPreference. Теперь используется через  use_case.

svetSanna committed:
Спринт 15.15. 1. Превратила EditSharedPreferencesUseCase в SharedPreferencesInteractor
2. Убрала из App прямое использование SharedPreference. Теперь используется через этот interactor.

svetSanna committed:
Спринт 15.16. Подправила строки.

svetSanna committed:
Спринт 15.16. 1. Исправила ошибку: убрала из domain jобращение к ресурсу (строке)

svetSanna committed:
Спринт 15.17. Подправила отбражение времени в MediaActivity

svetSanna commented:
Здравствуйте, постаралась исправить все замечания. Надеюсь, это получилось.) Спасибо за развернутые замечания! Немного смущает, что приходится передавать Activity в MediaPlayerRepositoryImpl для доступа к ресурсам, но, как сделать по-другому, не знаю. Комментарии пока не стала убирать, чтобы я могла понять, что я сделала.)

i-masloed commented on app/src/main/java/com/example/playlistmaker/domain/use_case/SharedPreferencesInteractor.kt:
```diff
@@ -0,0 +1,16 @@
+
+import com.example.playlistmaker.domain.repository.SharedPreferencesRepository
+
+class SharedPreferencesInteractor(private val repository: SharedPreferencesRepository) {
```
Рекомендация: интерактор- это отображение бизнес-функционала.  Например,  интерактор, используемы для настроек, должен содержать методы saveTheme(isDark:Boolean) и  isThemeDark():Boolean. В доменном слое не должно быть указано механизмов, как именно сохраняются данные
